app/src/main/java/app/nameplaceholder/sensorsapp/NotificationUpdater.kt

```diff
@@ -25,6 +25,7 @@ fun bindNotificationState(context: Context) {
             if (isActive) {
                 App.instance.repository.track()
             } else {
+                App.instance.repository.stop()
```

Comment:
This violates "Single responsibility principle". `NotificationUpdater` class should only be responsible for showing sensors data in notification when tracking is active.

Reply:
Yeah, but `NotificationUpdater` was already broken for that rule when the `App.instance.repository.track()` was placed here, don't you agree? I think a better approach could be to place both of these calls in the Application class or somewhere else and leave this class to just control the notifications.

Reply:
Not quite. Method naming is misleading. `repository.track()` doesn't start tracking. It return data stream (Kotlin Flow). 

In good old callback days this would be equal to something like:
```
repository.onUpdate((data) => { updateNotification(data) })
```

So this code only listens for change. While your added functionality manages outside resource (stops tracking).

Reply:
Maybe it not obvious from Kt syntax but it is:
```
if (isActive) {
    return App.instance.repository.track()
} else {
    return  emptyFlow()
}
```
while yours:
```
if (isActive) {
    return App.instance.repository.track()
} else {
    App.instance.repository.stop() <------- side effect
    return  emptyFlow()
}
```

Reply:
Ohh, you are right, the `stop()` call here is a side effect. But think about it, if it was the case that the Activity did not `collect`. The `App.instance.repository.track()` call here would be the responsible to start the tracking, right?
As far as I understand, the start of the tracking should be decoupled from from the Activity and from the notification management then.

Something like this in the Application would be a more independent solution, right?
```koltin
@UseExperimental(FlowPreview::class)
    fun manageTracking() {
        GlobalScope.launch {
            ForegroundService.isActive
                .filter { !it }
                .collect {
                    repository.stop()
                }
        }
        GlobalScope.launch {
            ForegroundService.isActive
                .filter { it }
                .switchMap { repository.track() }
                .collect {}
        }
    }
```